win-wasapi: Add support for capturing a process

jpark37 opened the pull request:
### Description
Use new Windows API for capturing audio from a process.

### Motivation and Context
Users want it.

### How Has This Been Tested?
Not extensively tested, but seems to work. Tested input/output/process, and fixed various issues.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

jpark37 commented:
Checks are probably busted until GitHub Actions provides 20348 SDK.

henke37 commented:
I wonder if this should be a mere separate mode or a separate source. Either way, should still keep it in the same dll.

As for the process selection, #2659 would greatly simplify it, right?

WizardCM commented:
So here are my thoughts on how I think this should be exposed to the user:

1. This should **not** be a thing you can set to a global audio source (the default Desktop Audio source, for example). This is because the core of that is modified in the Audio section of Settings, which a user shouldn't need to modify live. It is accessible from the Audio Mixer using the gear + Properties, but in this particular situation that should remain device-only.
2. The "Method" dropdown should have three options:
 a) Device
 b) OBS Capture Source
 c) Process / Application
3. For the "OBS Capture Source" method, the list should be populated with a list of Game Capture sources, a ---- separator, then a list of Window Capture sources. This should behave identically to Process mode, but doesn't require the user to manually specify a process, it just mirrors the Window / Game capture source's selected window.
4. For the Process / Application method, I agree that process ID would be bad. I'd say this method should essentially use the same window lookup/storage functions as Window Capture (same kind of dropdown). This is for when the user doesn't want to window capture the process but does want its audio (eg. a media player).

henke37 commented:
One thing for sure, I do not want duplicated code for the process/window selection.

Warchamp7 commented:
Have not tested my PR myself yet, but from a UX standpoint is there any reason this can't/shouldn't just be a checkbox in game capture / window capture that says "Capture audio directly" or something?

henke37 commented:
One UX improvement would be for the Game/Window capture source to have a "add audio capture source" button somewhere.

Warchamp7 commented:
> > Have not tested my PR myself yet, but from a UX standpoint is there any reason this can't/shouldn't just be a checkbox in game capture / window capture that says "Capture audio directly" or something?
> 
> What if you want to include some audio but don't want to show that window?
> 
> Sure you can hide the source but that's still taking unnecessary ressources.

A dropdown then to choose between capturing video, audio, or both perhaps?

Riotline commented:
Could work similarly to this plugin which does something similar but attaches before WASAPI since v2.0 I believe. It is only recent. [GitHub](https://github.com/bozbez/win-capture-audio/)

Warchamp7 commented:
> > > Have not tested my PR myself yet, but from a UX standpoint is there any reason this can't/shouldn't just be a checkbox in game capture / window capture that says "Capture audio directly" or something?
> > 
> > 
> > What if you want to include some audio but don't want to show that window?
> > Sure you can hide the source but that's still taking unnecessary ressources.
> 
> A dropdown then to choose between capturing video, audio, or both perhaps?
> 
> I do not think we should have two separate flows through the program for capturing a game or window depending on what the user wants to capture from it

I've thought on this more and changed my mind a bit. I'm mostly in agreement with what Matt suggested.

However, I don't think this should piggyback off the Audio Output Capture source type, at least UI/UX wise. Internally it can work that way if need be, but I think this should be an additional source type in the list and leave Audio Output Capture for devices only.

1. A setting in window and game capture properties to also grab audio from their specified process
2. An "Application Audio" source type that is purely audio capture

This way people can get video/audio from things like games all in one place, as well as intuitively grab audio only from things like Discord or a music app.

jpark37 commented:
I did a force-push, so you can select process by window now. I tested that Chrome and Spotify work. I'm a bit surprised that Chrome worked out of the box because I thought I was going to have to play shenanigans to find the audio process. I don't know how they map processes to windows/audio.

There's a lot of copy/paste from win-capture, but I didn't want to push that code into libobs until we get closer to deciding what to do. Window priority is not selectable and hard-coded to title to now.

Are we sure we want to add a new source type? I thought people wanted to consolidate them, e.g. merging window/game capture, but I might be remembering wrong.

RytoEX commented on plugins/win-wasapi/data/locale/en-US.ini:
```diff
@@ -3,3 +3,8 @@ AudioOutput="Audio Output Capture"
 UseDeviceTiming="Use Device Timestamps"
+Method="Method"
+Method.Device="Device"
+Method.WIndow="Process by window (Windows 10 2004 and up)"
```
Nitpick: "Method.Window"
(WI vs. Wi)

jpark37 replied:
Fixed.

RytoEX reviewed (changes requested):
Just nitpicking on a string name change.

I feel like there's a lot of code in here that is derived or copied entirely from code that we have elsewhere in the codebase for selecting application windows and determining information about a window.  Would it make sense to have these in some utils or platform code that plugins could use instead of copying them around from place to place?

RytoEX commented:
> There's a lot of copy/paste from win-capture, but I didn't want to push that code into libobs until we get closer to deciding what to do. Window priority is not selectable and hard-coded to title to now.

I'd personally prefer less copy/paste of code and making the promoting the window-helpers code out of win-capture to something more centrally accessible.  Though, that can wait till the UX guidance has firmed up, as you said.


> Are we sure we want to add a new source type? I thought people wanted to consolidate them, e.g. merging window/game capture, but I might be remembering wrong.

I _think_ what @Warchamp7 is suggesting here is to just register a source from the win-wasapi plugin that only offers "Window" or "Application" as the capture method instead of giving the user an option to select.  Ideally, Audio Output Capture would remain the same as it is now, and a new "Application Audio Capture" would be able to be pointed at an application and have it "just work".  We wouldn't have to teach users to use "Audio Output Capture" to capture application audio, and could just provide a straightforwardly named source type for it.  Presumably, this is easier for the user to understand and would make it harder to configure something incorrectly.

I think the idea of consolidation for window/game capture is less about removing a source type from the list for the sake of reduction and more about making the UX flow of "how do I capture an app" easier by providing one user-facing option so they don't have to think about the differences between "window" and "game" capture.  Though, that will likely require some automagic detection and a bunch of work, so I suspect it's a ways off, and I wouldn't let that idea get in the way of the idea here of having an extra simpler-to-configure source type with a straightforward name.

I'm not a mind reader though, so I'll let @Warchamp7 elaborate when he can.

jpark37 commented:
Added window priority if anyone wants to mess with that.

I don't mind adding a separate source type, but what UI options would it have? Window, window priority, and use device timestamps as currently?

Do we need modified window/game capture in this PR? I think the new source type by itself would be mergeable/shippable, and modifying the existing captures could come later. (And maybe by someone else if I run out of steam.)

jpark37 commented:
I split out the source type from "Audio Output Capture." as "Audio Process Output Capture." It won't be registered if Windows is older than 2004, but I'm not sure which version we actually need to gate on.

Part of the nastiness of moving the windows enumeration functionality from win-capture to libobs is that it locks the API in stone for external plugins to rely on, and the code could probably use some cleanup passes if we are going to make the functionality common and public.

jpark37 commented:
I added light/dark source icons for this by combining the window capture and audio output capture icons. It's bigger than the existing window capture icon, but I think it needs to be to see the detail.
![image](https://user-images.githubusercontent.com/10396506/133381165-be50eb9d-1c9a-41d8-ada2-75d3ab864acd.png)

jpark37 commented:
Removed draft status.

I moved windows-helpers.c/h and obfustcate.c/h from win-capture to libobs/util/windows. I noticed we didn't have existing documentation for other files in libobs/util/windows, so I didn't add any.

Warchamp7 commented:
Ryto understood me correctly. Leave Audio Output Capture as device only, add a separate entry specifically for process audio capture.

Does "Use device timestamps" make sense as a setting for this?

The window/game capture idea thingy is out of scope for this PR, just a relevant conversation I brought up for later UX improvements

jpark37 commented:
Yes, device timestamps still apply.

RytoEX reviewed (approved):
For style, this looks fine to me.  The functionality also works fine for me on Windows 10 21H1.  I'll give this my approval on those two points.

Still needs a UI/UX review from @Warchamp7.  Probably needs a review from @notr1ch for moving the window-helpers from win-capture to libobs.

jpark37 commented:
I'm thinking we should also switch away from adding an additional capture thread per audio capture, especially if we are going to allow them to be specified per-window. Windows has a built-in API for adding work to a common thread pool, which apparently is tuned for audio usage in terms of boosted priority and possibly other factors.
https://docs.microsoft.com/en-us/windows/win32/medfound/media-foundation-work-queue-and-threading-improvements

I have a prototype for using RTWQ, which the MF work queue seems to be mounted on top of. I just need to make sure it falls back gracefully to creating a capture thread if RTWQ support doesn't exist, e.g. Windows 7, before submitting a PR.

bozbez commented:
Rather than using the window selection logic from win-capture it might be better to enumerate the audio sessions directly using an `IMMDeviceEnumerator` and then an `IAudioSessionEnumerator` - you can get the session name and PID from `IAudioSessionControl2`. Would probably need also to get the executable name querying the PID (or maybe with information from the session identifiers? - don't know what they look like).

Aside from the selection, it would also be much better to use the `IAudioSessionNotification` interface (one per active device - needs integration with the current `IMMNotificationClient`) to update on session creation, and `IAudioSessionEvents` to update on session disconnection.

jpark37 commented:
Assuming `IAudioSessionEnumerator` is what the Volume Mixer uses, one drawback is that an application doesn't seem to appear in the list until a session is created, so I wouldn't replace window enumeration with it outright. We could add an alternate selection method down the line, but I think selection by window is fine for now.

I checked the "application missing from volume mixer until session start" scenario using the WindowsAudioSession sample:
https://github.com/microsoft/Windows-universal-samples/tree/main/Samples/WindowsAudioSession

A callback could be more reliable than IsWindow, which is susceptible to handle reuse by an unrelated window (although such an event seems very unlikely given its reliability in window capture), but I wouldn't want to use it if it fires just because the target application stopped their audio sessions for a moment because we might be late to reconnect if it restarted them, and we'd end up dropping samples. If the callback means that our loopback capture session will never receive audio again, e.g. process termination, that could be useful though.

bozbez commented:
Sure you won't see the application in the list until it opens it's session but I think sessions are opened pretty readily. On the other hand with the window selection logic you end up with a lot of bloat in the selection list, and you rely on the window process being either the audio output process or the parent of it. Either way it is probably fine to add that later, doesn't look too hard to implement.

`IAudioSessionEvents::OnStateChanged` gives you a `AudioSessionStateExpired` notification, which should be pretty much what we want. You could alternatively open the process handle and wait on that for the exit notification. I think the main advantage here is for the initiation of capture, since it would remove the need for polling.

jpark37 commented:
> Sure you won't see the application in the list until it opens it's session but I think sessions are opened pretty readily.

One scenario would be a media player that doesn't start a session until the user presses play. User opens their media player, looks for it in the OBS list, and then concludes that our application is dumb and broken because the media player isn't there.

> `IAudioSessionEvents::OnStateChanged` gives you a `AudioSessionStateExpired` notification, which should be pretty much what we want. You could alternatively open the process handle and wait on that for the exit notification.

I'm not sure AudioSessionStateExpired is strong enough; the application could still spin the audio systems back up and start playing, and we'd be late to reconnect. WaitForSingleProcess(hProcess, 0) is probably better than IsWindow though.

> I think the main advantage here is for the initiation of capture, since it would remove the need for polling.

On one hand, I hate polling. On the other hand, normal operation is a callback that occurs every 10 ms, so a 3 second period doesn't seem so bad in comparison.

IonicEcko commented:
I'm having the same issue with this that I have with the plugin that does the same thing (https://github.com/bozbez/win-capture-audio/releases/tag/v2.0.0-beta.3). The last few streams, about an hour in, the sound goes all distorted and horrible. It happens for a short time < 1 minute and then resolves and I've not had it reoccur.

The same issue has been reported on the plugin by a number of people so isn't limited to me however I think I'm the only one to test a build based on this commit.

Example of the sound is here -> https://clips.twitch.tv/ImpartialAmazonianLionCoolStoryBro-kQrRR8TK0c84oi2D

OS: Windows 11 Beta Branch
Single Audio Capture from Synth Riders (game exe)

Let me know if there is anything else I can do to assist.

jpark37 commented:
27.1 RC2 does not perform will on Windows 11 (unless you turn off all previews). 27.1 RC3 has a general performance fix for Windows 11, but since you mentioned other people are having issues on Windows 10, that probably isn't the main problem.

I am working on some changes to schedule the audio processing on an OS multimedia work queue instead of an ad-hoc thread, which will allegedly prioritize audio better, but who knows if it'll help the issue. Not sure when it'll be ready; I feel like I'm close, but I keep finding dumb mistakes and races in testing.

IonicEcko commented:
> 27.1 RC2 does not perform will on Windows 11 (unless you turn off all previews). 27.1 RC3 has a general performance fix for Windows 11, but since you mentioned other people are having issues on Windows 10, that probably isn't the main problem.
> 
> I am working on some changes to schedule the audio processing on an OS multimedia work queue instead of an ad-hoc thread, which will allegedly prioritize audio better, but who knows if it'll help the issue. Not sure when it'll be ready; I feel like I'm close, but I keep finding dumb mistakes and races in testing.

Yeah I made sure the preview was off for this. Happens regardless of preview.

jpark37 commented:
Updated the PR to use MMCSS "Capture" profile for audio capture, and MMCSS "Audio" profile for audio processing, which should give them much higher priority, and make less likely to miss processing deadlines. If that doesn't help with glitching, then I'm not sure what to try next.

IonicEcko commented:
> Updated the PR to use MMCSS "Capture" profile for audio capture, and MMCSS "Audio" profile for audio processing, which should give them much higher priority, and make them less likely to miss processing deadlines. If that doesn't help with glitching, then I'm not sure what to try next.

Happy to give this a shot tomorrow. Does this auto build the app or would I have to work out how to build it?

RytoEX commented:
> 
> 
> > Updated the PR to use MMCSS "Capture" profile for audio capture, and MMCSS "Audio" profile for audio processing, which should give them much higher priority, and make them less likely to miss processing deadlines. If that doesn't help with glitching, then I'm not sure what to try next.
> 
> Happy to give this a shot tomorrow. Does this auto build the app or would I have to work out how to build it?

Build artifacts are available under the "Checks" tab in the GitHub UI under "Artifacts".

Windows 64-bit:  https://github.com/obsproject/obs-studio/suites/3880300208/artifacts/96603950

bozbez commented:
> One scenario would be a media player that doesn't start a session until the user presses play. User opens their media player, looks for it in the OBS list, and then concludes that our application is dumb and broken because the media player isn't there.

I think that is just a question of presentation though; of course this would happen if the selection drop-down is named "Window" but if it was named something like "Audio Stream" and maybe even with a small text below it "Only applications that are currently playing audio will appear in the list" you avoid this.

> I'm not sure AudioSessionStateExpired is strong enough; the application could still spin the audio systems back up and start playing, and we'd be late to reconnect. WaitForSingleProcess(hProcess, 0) is probably better than IsWindow though.

It seems unlikely (although I haven't confirmed) that you would be able to take an audio session from expired back to active, which means if the application did create another session then you would process it immediately as a new session notification, so I don't think you would be late at all. With the current polling, we are already up to 3s late for the more common case of capturing audio from a newly started application.

jpark37 commented:
Pushed a Windows 7 fix (even though we don't officially support 7). Apparently, the callback event isn't signaled for loopback on that OS.

> we are already up to 3s late for the more common case of capturing audio from a newly started application.

The scenario I'm wondering about is if you press stop on a media player, and it fires an expired notification, which we respond to by tearing things down. Then you press play, and the new session notification comes in, and we would be late by some small amount to spin things back up. If we ignore the expired/new notifications and leave the client active, maybe we wouldn't lose any samples at all.

jpark37 commented:
I just streamed a bit ago with latest changes, and reproduced the choppy audio about an hour in, and it resolved itself after about a minute. One more thing to try is to uncheck "Use Device Timestamps" because they might be bad, or OBS isn't synchronizing to them correctly.

IonicEcko commented:
I'll give that a go, am about to start myself so can test it now for you. But yeah exactly the same symptoms for me, happens about an hour in and resolves shortly after.

IonicEcko commented:
I don't have an update for you sorry. I unticked the Device Timestamps box but it seemed to introduce a large amount (500+ms) of latency into the audio capture which is very obvious in a rhythm game. I fixed it up once with an audio delay but then it hitched later in the stream and the delay doubled so I had to tick the option again and restart the hook.

Will try to give it another go next time.

jpark37 commented:
Assuming the repro is consistent, I can try to log timestamps and AUDCLNT_BUFFERFLAGS states, and see if something weird happens during the glitchy parts. Sucks if each test takes over an hour, but I guess that's not as bad as multiple hours.

IonicEcko commented:
If I can help in any way please let me know.

jpark37 commented:
I haven't attempted to diagnose the problem beyond noticing that OBS should probably handle AUDCLNT_BUFFERFLAGS_DATA_DISCONTINUITY more smartly, but I pushed a hope-fix that temporarily ignores the timestamp checkbox setting, and generates a fake timestamp based on the number of processed frames. Haven't done much testing beyond a 5 second recording sanity check, but what the heck.

bozbez commented:
> The scenario I'm wondering about is if you press stop on a media player, and it fires an expired notification, which we respond to by tearing things down. Then you press play, and the new session notification comes in, and we would be late by some small amount to spin things back up. If we ignore the expired/new notifications and leave the client active, maybe we wouldn't lose any samples at all.

Do you have an example of a media player that actually tears down it's session each time you pause? The common ones I tried just keep the session open. I'd also expect the spin-up time to be pretty minimal here, but no way to try without a test application.

As a side note, two capture sources on the same window won't work. It seems that whatever is happening for the audio routing is done with PIDs, so attempting a second session from the same capture process just gives you an audio capture client that never fires. Not much to do about this apart from run the capture in a separate process.

YuvalShor commented:
When trying to test the latest artifact, I've noticed I don't have Twitch/YouTube integration anymore, is that to be expected?
![image](https://user-images.githubusercontent.com/10600102/135634895-8855a0a0-3930-4105-aa1c-cc15e7c158bc.png)

RytoEX commented:
> When trying to test the latest artifact, I've noticed I don't have Twitch/YouTube integration anymore, is that to be expected?

Yes, this is expected. PR builds do not have access to GitHub secrets/tokens, which are used to enable service integrations like Twitch and YouTube.

YuvalShor commented:
> > When trying to test the latest artifact, I've noticed I don't have Twitch/YouTube integration anymore, is that to be expected?
> 
> Yes, this is expected. PR builds do not have access to GitHub secrets/tokens, which are used to enable service integrations like Twitch and YouTube.

OK, thanks for clarifying.
Was just making sure.

jpark37 commented:
I tested the hope-fix today for faking the timestamps, and it might work. I checked a 4h 45m recording with 10 second skips, and didn't find any glitchy parts, and was still reasonably synced up to gameplay throughout.

> Do you have an example of a media player that actually tears down it's session each time you pause?

This sample might since it seems to spin up on play, but I haven't checked if it tears down on stop:
https://github.com/microsoft/Windows-universal-samples/tree/main/Samples/WindowsAudioSession

> As a side note, two capture sources on the same window won't work. It seems that whatever is happening for the audio routing is done with PIDs, so attempting a second session from the same capture process just gives you an audio capture client that never fires. Not much to do about this apart from run the capture in a separate process.

Why would a user want to capture the same process more than once? That doesn't seem like a desirable feature.

jpark37 commented:
Set as draft. Builds on #5369.

IonicEcko commented:
Unfortunately it appears to still be a problem, at roughly the same time. It happened about 47 minutes into stream but I had OBS running longer than usual before stream. Checked the logs and there doesn't seem to be any entries around the time it happened.

Same experience as before using the a4d7f1f4c compile (most recent as at time of go live - 2021-10-04-a4d7f1f4c-27.1.3-win64.zip)

I do notice a lot of WASAPI: Device '' failed to start which might be an audio cap on another source or something that wasn't active. Is there any way it can treat it like the game cap where it doesn't spam try?

jpark37 commented:
Empty string is an application audio capture. If the application isn't open, it'll spam that. I'm on the fence about removing it; it only shows up once every three seconds, and support may need it. I can improve the logging to make it show something more useful like the exe|title|window_class or something.

I can look into putting together a logging build that will write the virtual timestamp, real timestamp, frame count, and WASAPI flags for every callback to see if there's some pattern that we could do something about. Hopefully that's not too much logging to throw off the timing or fill up disk too quickly.

jpark37 commented:
Fwiw, I did another 2 hour session yesterday, and both recording and wasapi_log we're clean. It'll be interesting to see what your wasapi_log ends up looking like.

I'm also bad at estimating; the two hour session generated less than 30 MB. Be sure to close OBS gracefully so fclose gets called and flushes the log completely to disk.

jpark37 commented:
Unfortunately, the log looks pretty clean.

I wonder if you need a more recent version of Windows? The API is usable starting with Windows 10 2004 (19041), but is only documented to work for 20348, which is Windows Server 2022. Windows 11 is the first consumer OS at 22000.

My clean runs have been on Windows 11.

IonicEcko commented:
I'm on 2247.1000

WizardCM commented:
I found a minor UX bug that could confuse users - each Microsoft Edge tab is treated as a separate audio source, but it's categorized under `explorer.exe`. I realise this is likely limited to MS Edge, but thought I'd at least mention it.

![image](https://user-images.githubusercontent.com/941350/139561074-3de25f57-ab86-4018-b0b9-44c044310da0.png)

bozbez commented:
It seems that Valorant and Call of Duty: Warzone output audio for in-game voice chat through a process separate to the process tree of the game window, while the rest of in-game audio does work.

Not my personal issue though, so don't know if that is because of a different audio session or some cross-process shared session which I think is a thing?

jpark37 commented:
To make it clear to anyone who's interested, I'm not planning to remove draft status until the funky audio hiccup is resolved, but I also don't have any insight into what's causing it, or know anyone at MS who can look into it. I expect this PR will be stuck for a long while. Obviously, you can use bozbez's plugin in the meantime.

shleeable commented:
This rules. Thank you.

RytoEX commented:
> I'm having the same issue with this that I have with the plugin that does the same thing (https://github.com/bozbez/win-capture-audio/releases/tag/v2.0.0-beta.3). The last few streams, about an hour in, the sound goes all distorted and horrible. It happens for a short time (< 1 minute) and then resolves and I've not had it reoccur again in the same stream (but none have been uber long).
> 
> The same issue has been reported on the plugin by a number of people so isn't limited to me however I think I'm the only one to test a build based on this commit.
> 
> Example of the sound is here -> https://clips.twitch.tv/ImpartialAmazonianLionCoolStoryBro-kQrRR8TK0c84oi2D
> 
> OS: Windows 11 Beta Branch Single Audio Capture from Synth Riders (game exe)
> 
> Let me know if there is anything else I can do to assist.
> 
> Edit: OBS Log, it won't show the stream because I use NDI to send it to another PC. (https://obsproject.com/logs/hfFw22FC9YLidNVw)
> 
> Edit2 (Last one sorry): I tried to work out any commonality in setup with the other person having the trouble on the plugins Discord and the only common factor seemed to be the Ryzen 5900x CPU... Everything else differed, they were on Win 10, didn't use NDI etc.

> 1 hour after OBS is started it happens, gets progressively worse over 10-20 seconds and then progressively better over the next little bit until its gone and then never happens again (at least not within a ~4 hour window).

> I use the default 48khz sample rate across my sound setup + OBS, at a 24-bit rate but the VR headset (which the processes audio would be targetting - if that even matters) is only compatible with 16-bit.

@IonicEcko 
I wanted to follow-up here and try to gather some more info.  You said that you also experienced the same issue in [the win-capture-audio plugin](https://github.com/bozbez/win-capture-audio/releases/tag/v2.0.0-beta.3) and that the issue was reported there.

1. Could you point me to the GitHub Issue for the "win-capture-audio" plugin so that we're on the same page?  If others have experienced the same issue, it might be helpful if they were available to help test these builds.
2. Does this affect the entire audio track output, or does it only affect the application/process-specific audio?
3. Does this happen 1 hour after opening OBS, or one hour into a streaming/recording session?
4. Does this also happen in local recordings?
5. Where are you checking the audio bit depth of the applications?


@jpark37 
Is there a Microsoft sample app that affected users can try to reproduce this with?

jpark37 commented:
> Is there a Microsoft sample app that affected users can try to reproduce this with?

I was playing Fall Guys when I reproduced the one time. It hasn't happened for me since.

IonicEcko commented:
@RytoEX 

> Could you point me to the GitHub Issue for the "win-capture-audio" plugin so that we're on the same page? If others have experienced the same issue, it might be helpful if they were available to help test these builds.

It was all based on discussions on their Discord I believe. [This](https://github.com/bozbez/win-capture-audio/issues/31) was the issue that I originally commented on though.

> Does this affect the entire audio track output, or does it only affect the application/process-specific audio?

Only seems to affect the audio being captured by this plugin/feature.

> Does this happen 1 hour after opening OBS, or one hour into a streaming/recording session?

One hour post hook. For example if I change the AAC source to capture a different EXE and back again, the 1 hour (ish - very close to) starts again.

> Does this also happen in local recordings?

Unsure. I believe so but can't say that with certainty because I have a horrible memory. I currently have a broken stream setup but as soon as its fixed I can try to replicate again.

> Is the bad audio noticeable in audio monitoring?

Unknown, because it happens while I'm playing VR I haven't been able to test this.

> Where are you checking the audio bit depth of the applications?

I was just checking the audio properties for the different inputs and outputs in Windows.

IonicEcko commented:
I just wanted to check back in to say I built a new 12th gen based build that still experiences the issue so semi-rules out hardware causes.

The new PC is on completely new hardware with fresh Windows install. Still happening about 60 minutes in to every stream.

Is there a feedback post for this I can help upvote to see if we can get some traction from the MSFT side?

jpark37 commented:
@Warchamp7 Uhh... What are you doing?

Warchamp7 commented:
> @Warchamp7 Uhh... What are you doing?

I have made a terrible mistake I am sorry

EddyGharbi commented:
I found an issue while using Spotify. If I'm listening to a specific song while I capture the app in OBS, it will show up as `Nelly - Hot In Herre:Chrome_WidgetWin_0:Spotify.exe` for example. If, when I restart OBS, Spotify happens to be playing another song, it will cause OBS to not pickup Spotify anymore but the next app on the list.

As a user I would expect that the fact I was capturing Spotify is saved rather than the song associated to it. What do you guys think?

jpark37 commented:
If someone is still able to reproduce, we could use an ETL trace by following these steps:

- From an admin command prompt: `wpr -start obs.wprp`
- Start the scenario – keep capturing until the problem occurs
- `wpr -stop repro.etl`
- Provide repro.etl to us privately since it may contain personal information. You can find us in Discord.
- Optional: Provide audio from the recording. I don't know how to separate from a video recording off the top of my head, but I'm sure Google can help.

Get build here (probably want obs-win64): https://github.com/obsproject/obs-studio/actions/runs/2000843739
Get obs.wprp here: [obs.wprp.zip](https://github.com/obsproject/obs-studio/files/8292628/obs.wprp.zip)

mufunyo commented:
> * Optional: Provide audio from the recording. I don't know how to separate from a video recording off the top of my head, but I'm sure Google can help.

You can use [gMKVExtractGUI](https://sourceforge.net/projects/gmkvextractgui/files/latest/download) (requires [MKVToolNix](https://www.fosshub.com/MKVToolNix.html)) to extract the audio from an OBS recording. Assuming default OBS settings (160kbit AAC audio) the file should not be very large.

jpark37 commented:
Decided to merge the changes while marking the source as "(BETA)" until someone can help us repro, and solve the issue. Draft status removed. Hopefully we can get this resolved during the prerelease cycles for v28.

jpark37 commented:
I copy/pasted in the code that window capture uses, and pushed it. The Spotify steps seem to work for me now.

jpark37 commented:
This is also not the same issue that EddyGharbi raised, which is complicated by the fact that Spotify uses a Chrome-derived window class name. We don't want Chrome/Discord to be accidentally seen/heard if Spotify isn't around.

In theory, the match by executable option should work for his scenario since "Spotify.exe" should be unique enough, but I see the code still rejects this because of the Chrome class name. I feel like that's not how this is supposed to work, but I'd have to ask around.

jpark37 commented:
It seems restoring executable name matching is probably doable, at least for Spotify. It was disabled for UWP apps because a lot of them were running a "ApplicationHost.exe", so the exe name was meaningless, but Chrome apps don't seem to have this problem, so we should treat them differently. We've been bouncing around some ideas.

I kind of hate the source toolbar tbh, but people are used to it, and I understand the workflow improvement. I'll see about aligning with window capture there.

jpark37 commented:
Pushes updates to allow Spotify.exe matching, and show Window in the source toolbar.

style1925 commented:
1, Applications that do not appear in the choices
"Video.UI.exe" does not appear in the choices.
Is this a specification?
It does not appear in the window capture either. (Is it intentionally excluded?)

2, Unintended audio capture
*I found this problem when I made a mistake in specifying the application.

Steps to reproduce
1, Start "Battle.net"
2, Start "Overwatch"
3, Specify "Battle.net.exe" in ”Application Audio Capture (BETA)"
4, Even if "Battle.net.exe" is specified, the audio of "Overwatch" will be captured.

Video recorded from step 3
https://youtu.be/Uc0BH-7CMKw

**Anything else we should know?**
I used a translation. Sorry if the English is not correct.

style1925 commented:
Thank you for letting me know!

**Is the following behavior intentional?**
I discovered this when I opened the Log Viewer and was looking at the logs.

Selecting a ”Application Audio Capture (BETA)" will log "update settings" in the log.
In the video, the two sources are selected alternately for clarity.
https://youtu.be/brwHcjlLgUE

jpark37 commented:
> Selecting a ”Application Audio Capture (BETA)" will log "update settings" in the log.

I updated the logging behavior to match window capture.

Warchamp7 commented:
**Major Concerns**
It is not possible to audio capture an app twice. Any hooks trying to capture after the first will silently fail.

Is it possible to prevent a user from picking the same process twice, or displaying a red error on the selection or anything if they do?

**Minor Concerns**:

There is a brief cut/crackle when using a Cut transition between two scenes that both have the same audio capture.
https://user-images.githubusercontent.com/1554753/175801059-95a01a2c-fb50-4504-8de6-cce8d10cb91e.mp4

jpark37 commented:
> Is it possible to prevent a user from picking the same process twice, or displaying a red error on the selection or anything if they do?

The audio mixer should show no audio getting through. It's similar to the video capture sources; if it doesn't work, you just get no picture in the scene preview, and we log the failure.

> There is a brief cut/crackle when using a Cut transition between two scenes that both have the same audio capture.

Are you using the same source, or two separate sources pointed at the same process? Is this an issue with regular Audio Output Capture?

jpark37 commented:
Fwiw, I can only reproduce the scene cut issue if I use separate sources. I don't hear a cut if I add the existing source to the other scene.

Warchamp7 commented:
> > Is it possible to prevent a user from picking the same process twice, or displaying a red error on the selection or anything if they do?
> 
> The audio mixer should show no audio getting through. It's similar to the video capture sources; if it doesn't work, you just get no picture in the scene preview, and we log the failure.

The mixer shows no audio, but it's not really clear to the user why unless they know they have another source trying to hook the same process.

I am aware game capture suffers the same problem, but I'm not a fan of that either :P

> > There is a brief cut/crackle when using a Cut transition between two scenes that both have the same audio capture.
> 
> Are you using the same source, or two separate sources pointed at the same process? Is this an issue with regular Audio Output Capture?

Same source in both scenes, as I cannot get two sources to capture the same process.

jpark37 commented:
Probably needs a follow-up PR for Yami theme.

micsthepick commented:
will this support surround sound?

z0fa commented:
is it possible to add an hotkey to capture the current focused window/game like in the gamecapture source?

SnorreSelmer commented:
> will this support surround sound?

If the application outputs surround sounds, then OBS should be able to pull surround sound. To the API in question, it's just data.

pandages commented:
I'm having the Noisy after 30-ish minutes issue with this audio capture method.  Are you still looking for logs?

RytoEX commented:
> I'm having the Noisy after 30-ish minutes issue with this audio capture method. Are you still looking for logs?

We are still collecting information on this, yes. I can be reached directly on Discord at RytoEX#5545. Specifically, we are still looking for the data outlined [here](https://github.com/obsproject/obs-studio/pull/5218#issuecomment-1071522440).

Destroy666x commented:
This feature is unfortunately still lacking, compared to an available win-capture-audio plugin, which has exclusions. Please consider adding that. So that e.g. few music apps and all non-music apps can be easily configured into 2 separate sources.

Nipeno commented:
Wish this had the exclusion feature from win-audio-capture.
